In this example, we will see GEPA evolve the whole DSPy program (not just the instruction), including modifying the structure/dataflow of the program. We will use GEPA to tune a simple dspy.ChainOfThought module for MATH questions into a full DSPy program.

In [1]:
# Install python-dotenv and load .env.dev if present
try:
    from dotenv import load_dotenv
    import os
    if os.path.exists(".env.dev"):
        load_dotenv(".env.dev")
except ImportError:
    print("python-dotenv is not installed. Skipping .env.dev loading.")


In [2]:
import dspy

In [3]:
import random

from dspy.datasets import MATH

dataset = MATH(subset="algebra")

# Shuffle the train and dev sets
random.Random(0).shuffle(dataset.train)
random.Random(0).shuffle(dataset.dev)

print(len(dataset.train), len(dataset.dev), len(dataset.test))

/home/jacobjensen/Projects/prompt_evolution/gepa/.venv/lib/python3.11/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


350 350 487


Let's inspect an example from the training set.

In [4]:
example = dataset.train[0]
print("Question:", example.question)
print("Answer:", example.answer)

Question: If $f(x) = x^3 - 6x^2 + 3x - 4$, $g(x) = x^3 + 5x^2 + 9x - 2$, then find the constant term of $f(g(x))$.
Answer: -42


Let's define a simple DSPy program to solve this task.

Unlike dspy.GEPA that can take an instantiated DSPy module as input, here, we want to evolve the full DSPy program. Hence, a candidate here is the source code as string. The seed program does not need to be sophisticated, it just needs to demonstrate what the expected input/output interface is, and possibly the available tools. You can also include any additional information about the environment as a comment.

In [5]:
program_src = """import dspy
program = dspy.ChainOfThought("question -> answer")"""

GEPA interfaces with external frameworks through an adapter. In this case, we integrate GEPA with a DspyAdapter.

In [6]:
# 1) See which Python your kernel is using and the import path
import sys, platform
print("Python:", sys.version)
print("Executable:", sys.executable)
print("First 5 sys.path entries:")
print("\n".join(sys.path[:5]))

# 2) Check whether Python can even see the module chain you want
import importlib.util as ilu

def show_spec(name):
    spec = ilu.find_spec(name)
    print(f"{name:60} -> {'FOUND' if spec else 'NOT FOUND'}")
    if spec:
        print("  origin:", spec.origin)
        print("  submodule_search_locations:", spec.submodule_search_locations)

for m in [
    "gepa",
    "gepa.adapters",
    "gepa.adapters.dspy_full_program_adapter",
    "gepa.adapters.dspy_full_program_adapter.full_program_adapter",
    "dspy",
]:
    show_spec(m)

Python: 3.11.11 (main, Feb 12 2025, 14:51:05) [Clang 19.1.6 ]
Executable: /home/jacobjensen/Projects/prompt_evolution/gepa/.venv/bin/python
First 5 sys.path entries:
/home/jacobjensen/.local/share/uv/python/cpython-3.11.11-linux-x86_64-gnu/lib/python311.zip
/home/jacobjensen/.local/share/uv/python/cpython-3.11.11-linux-x86_64-gnu/lib/python3.11
/home/jacobjensen/.local/share/uv/python/cpython-3.11.11-linux-x86_64-gnu/lib/python3.11/lib-dynload

/home/jacobjensen/Projects/prompt_evolution/gepa/.venv/lib/python3.11/site-packages
gepa                                                         -> FOUND
  origin: /home/jacobjensen/Projects/prompt_evolution/gepa/src/gepa/__init__.py
  submodule_search_locations: ['/home/jacobjensen/Projects/prompt_evolution/gepa/src/gepa']
gepa.adapters                                                -> FOUND
  origin: /home/jacobjensen/Projects/prompt_evolution/gepa/src/gepa/adapters/__init__.py
  submodule_search_locations: ['/home/jacobjensen/Projects/prompt_

In [7]:


# 3) Enumerate what's inside gepa.adapters (using pkgutil)
import pkgutil, importlib, sys

gepa_adapters = importlib.import_module("gepa.adapters")
print("gepa.adapters.__path__ ->", list(gepa_adapters.__path__))

print("Subpackages/modules under gepa.adapters:")
for mod in pkgutil.iter_modules(gepa_adapters.__path__):
    print(" -", mod.name, "(package)" if mod.ispkg else "(module)")

gepa.adapters.__path__ -> ['/home/jacobjensen/Projects/prompt_evolution/gepa/src/gepa/adapters']
Subpackages/modules under gepa.adapters:
 - anymaths_adapter (package)
 - default_adapter (package)
 - dspy_adapter (package)
 - dspy_full_program_adapter (package)
 - ifbench_adapter (package)
 - terminal_bench_adapter (package)


In [8]:
from gepa.adapters.dspy_full_program_adapter.full_program_adapter import DspyAdapter

In [9]:
def metric_fn(example, pred, trace=None):
    score = dataset.metric(example, pred)
    if score:
        feedback_text = f"The provided answer '{pred.answer}' is correct."
    else:
        feedback_text = f"The provided answer '{pred.answer}' is incorrect. The correct answer is '{example.answer}'. Here's the step by step solution:\n{example.reasoning}"
    return dspy.Prediction(score=score, feedback=feedback_text)

In [10]:
reflection_lm = dspy.LM(model="openai/gpt-4.1", max_tokens=32000)  # temperature=1
adapter = DspyAdapter(
    task_lm=dspy.LM(model="openai/gpt-4.1-nano", max_tokens=32000),
    metric_fn=metric_fn,
    num_threads=80,
    reflection_lm=lambda x: reflection_lm(x)[0],
)

Let's evaluate the base program

In [11]:
o = adapter.evaluate(dataset.test, {"program": program_src})

2025/09/13 16:34:54 INFO dspy.evaluate.evaluate: Average Metric: 328.0 / 487 (67.4%)


The base program obtains a score of 67.1%

Let's launch the GEPA optimization.

In [12]:
from gepa import optimize

o = optimize(
    seed_candidate={"program": program_src},
    trainset=dataset.train,
    valset=dataset.dev[:200],
    adapter=adapter,
    reflection_lm=lambda x: reflection_lm(x)[0],
    max_metric_calls=2000,
    display_progress_bar=True,
)

GEPA Optimization:  10%|█         | 200/2000 [00:21<03:13,  9.30rollouts/s]

Iteration 0: Base program full valset score: 0.66
Iteration 1: Selected program 0 score: 0.66
Average Metric: 2.00 / 3 (66.7%): 100%|██████████| 3/3 [00:04<00:00,  1.61s/it]

2025/09/13 16:35:20 INFO dspy.evaluate.evaluate: Average Metric: 2.0 / 3 (66.7%)



Iteration 1: Proposed new text for program: import dspy
from typing import Literal

class MathQA(dspy.Signature):
    """
    Solve the given math word problem step by step, showing all necessary reasoning and calculations.
    Then, extract and return ONLY the final numeric answer as a string, with no units or extra text.
    
    Instructions:
    - Show clear, step-by-step reasoning in the 'reasoning' field.
    - In the 'answer' field, provide ONLY the final numeric answer (e.g., '42', '6', '32'), not a sentence or explanation.
    - Do not include units, words, or extra formatting in the answer.
    - If there are multiple valid numeric answers, sum them and return the sum as a string.
    - Common pitfalls: Do not restate the question or include sentences in the answer field. Do not include units or box the answer.
    - Example:
        Question: "If x + 2 = 5, what is x?"
        Reasoning: "x + 2 = 5 => x = 5 - 2 = 3"
        Answer: "3"
    - Example:
        Question: "Let 

2025/09/13 16:35:35 INFO dspy.evaluate.evaluate: Average Metric: 3.0 / 3 (100.0%)
2025/09/13 16:37:58 INFO dspy.evaluate.evaluate: Average Metric: 155.0 / 200 (77.5%)
GEPA Optimization:  20%|██        | 406/2000 [03:04<13:39,  1.94rollouts/s]

Iteration 1: New program is on the linear pareto front
Iteration 1: Full valset score for new program: 0.775
Iteration 1: Full train_val score for new program: 0.775
Iteration 1: Individual valset scores for new program: [True, True, True, False, False, False, True, True, True, True, True, True, True, True, False, True, True, True, True, True, True, True, True, True, False, True, True, True, True, True, True, False, True, True, True, True, True, True, True, False, True, False, True, True, True, True, False, True, True, True, True, True, True, False, True, True, True, True, True, True, True, False, True, False, True, True, True, True, True, True, True, True, True, True, True, True, True, False, True, False, False, True, False, True, True, True, False, True, True, True, False, True, True, True, True, True, True, True, False, True, True, False, True, True, True, True, True, True, False, False, False, False, True, True, True, False, True, True, True, True, True, False, True, True, False, T

2025/09/13 16:38:01 INFO dspy.evaluate.evaluate: Average Metric: 0.0 / 3 (0.0%)



Iteration 2: Proposed new text for program: import dspy
import re
from typing import Literal, Optional

class MathQA(dspy.Signature):
    """
    Solve the given math word problem step by step, showing all necessary reasoning and calculations.
    Then, extract and return ONLY the final answer in the correct format, according to the problem type.

    Instructions:
    - Show clear, step-by-step reasoning in the 'reasoning' field.
    - In the 'answer' field, provide ONLY the final answer, formatted as required:
        * If the answer is a single number, return just the number (e.g., '42', '6', '32'), no units or extra text.
        * If the answer is an interval or union of intervals, return in standard interval notation (e.g., '[0,1)', '(-∞,2] ∪ (3,5)').
        * If the answer is a polynomial or algebraic expression, return the fully simplified expression (e.g., '6x^2 + 30x + 36').
        * If the answer must be rounded, follow the problem's rounding instructions exactly.
    - D

2025/09/13 16:38:30 INFO dspy.evaluate.evaluate: Average Metric: 0.0 / 3 (0.0%)
GEPA Optimization:  21%|██        | 412/2000 [03:36<16:40,  1.59rollouts/s]

Iteration 2: New subsample score is not better, skipping
Iteration 3: Selected program 0 score: 0.66
Average Metric: 3.00 / 3 (100.0%): 100%|██████████| 3/3 [00:04<00:00,  1.34s/it]

2025/09/13 16:38:34 INFO dspy.evaluate.evaluate: Average Metric: 3.0 / 3 (100.0%)
GEPA Optimization:  21%|██        | 415/2000 [03:40<16:57,  1.56rollouts/s]


Iteration 3: All subsample scores perfect. Skipping.
Iteration 3: Reflective mutation did not propose a new candidate
Iteration 4: Selected program 1 score: 0.775
Average Metric: 2.00 / 3 (66.7%): 100%|██████████| 3/3 [00:04<00:00,  1.64s/it] 

2025/09/13 16:38:39 INFO dspy.evaluate.evaluate: Average Metric: 2.0 / 3 (66.7%)



Iteration 4: Proposed new text for program: import dspy

class MathQA(dspy.Signature):
    """
    Solve the given math word problem step by step, showing all necessary reasoning and calculations.
    Then, extract and return ONLY the final numeric answer as a string, with no units or extra text.

    Instructions:
    - Show clear, step-by-step reasoning in the 'reasoning' field.
    - In the 'answer' field, provide ONLY the final numeric answer (e.g., '42', '6', '32', '3/2', '1.5'), not a sentence or explanation.
    - If the answer is a fraction, return it in reduced form (e.g., '3/2'), unless a decimal is more natural (e.g., '1.5' if the problem expects decimals).
    - Do not include units, words, or extra formatting in the answer.
    - If the answer is boxed, in LaTeX, or appears in a sentence, extract just the number or fraction.
    - If there are multiple valid numeric answers, sum them and return the sum as a string.
    - Never sum or concatenate digits from the reasoning;

2025/09/13 16:38:58 INFO dspy.evaluate.evaluate: Average Metric: 2.0 / 3 (66.7%)
GEPA Optimization:  21%|██        | 421/2000 [04:03<20:56,  1.26rollouts/s]

Iteration 4: New subsample score is not better, skipping
Iteration 5: Selected program 1 score: 0.775
Average Metric: 3.00 / 3 (100.0%): 100%|██████████| 3/3 [00:07<00:00,  2.55s/it]

2025/09/13 16:39:05 INFO dspy.evaluate.evaluate: Average Metric: 3.0 / 3 (100.0%)
GEPA Optimization:  21%|██        | 424/2000 [04:11<22:23,  1.17rollouts/s]


Iteration 5: All subsample scores perfect. Skipping.
Iteration 5: Reflective mutation did not propose a new candidate
Iteration 6: Selected program 0 score: 0.66
Average Metric: 2.00 / 3 (66.7%): 100%|██████████| 3/3 [00:06<00:00,  2.19s/it] 

2025/09/13 16:39:12 INFO dspy.evaluate.evaluate: Average Metric: 2.0 / 3 (66.7%)



Iteration 6: Proposed new text for program: import dspy
from typing import Optional

class MathSolutionSignature(dspy.Signature):
    """
    Solve the given math problem step by step, showing all necessary reasoning and calculations.
    - Provide a clear, detailed solution in the 'solution' field, using LaTeX for mathematical expressions.
    - In the 'answer' field, give only the final answer, formatted as simply and precisely as possible.
    - If the answer is a fraction, always use LaTeX (e.g., '\\frac{a}{b}').
    - Avoid common pitfalls: do not skip steps, always check for extraneous solutions, and ensure the answer matches the problem's requirements (e.g., domain restrictions, inverse function existence).
    - For graph or function problems, explicitly state any assumptions and show all algebraic manipulations.
    - For repeating decimals, always convert to the simplest fraction and use LaTeX.
    - If multiple answers are possible, list all valid ones in the answer field, 

2025/09/13 16:39:25 WARNING dspy.clients.lm: LM response was truncated due to exceeding max_tokens=32000. You can inspect the latest LM interactions with `dspy.inspect_history()`. To avoid truncation, consider passing a larger max_tokens when setting up dspy.LM. You may also consider increasing the temperature (currently 0.0)  if the reason for truncation is repetition.
2025/09/13 16:39:28 INFO dspy.evaluate.evaluate: Average Metric: 0.0 / 3 (0.0%)
GEPA Optimization:  22%|██▏       | 430/2000 [04:34<28:53,  1.10s/rollouts]

Iteration 6: New subsample score is not better, skipping
Iteration 7: Selected program 0 score: 0.66
Average Metric: 2.00 / 3 (66.7%): 100%|██████████| 3/3 [00:03<00:00,  1.27s/it] 

2025/09/13 16:39:32 INFO dspy.evaluate.evaluate: Average Metric: 2.0 / 3 (66.7%)



Iteration 7: Proposed new text for program: import dspy
import re

class MathQA(dspy.Signature):
    """
    Solve the given math problem step by step, showing all intermediate calculations and reasoning.
    - Carefully check each arithmetic step for accuracy.
    - For numeric answers, ensure correct formatting (e.g., no commas in numbers unless specified, round as instructed).
    - For coordinate answers, use the format (x, y) with no spaces after commas.
    - For decimal answers, round as specified in the question.
    - If the question requests an exact value, do not round.
    - Common pitfalls: arithmetic errors, misreading instructions, incorrect answer formatting.
    - Successful strategy: Write out all steps, double-check calculations, and clearly state the final answer in the required format.
    """
    question: str = dspy.InputField(desc="A math problem to solve")
    reasoning: str = dspy.OutputField(desc="Step-by-step solution with all calculations and logic")
    a

2025/09/13 16:39:50 INFO dspy.evaluate.evaluate: Average Metric: 2.0 / 3 (66.7%)
GEPA Optimization:  22%|██▏       | 436/2000 [04:56<35:43,  1.37s/rollouts]

Iteration 7: New subsample score is not better, skipping
Iteration 8: Selected program 0 score: 0.66
Average Metric: 2.00 / 3 (66.7%): 100%|██████████| 3/3 [00:03<00:00,  1.28s/it]

2025/09/13 16:39:54 INFO dspy.evaluate.evaluate: Average Metric: 2.0 / 3 (66.7%)



Iteration 8: Proposed new text for program: import dspy
import re

class MathQA(dspy.Signature):
    """
    Solve the given math problem step by step, showing all necessary reasoning, calculations, and checks (such as domain restrictions or extraneous solutions).
    - For answers involving radicals or fractions, always simplify fully.
    - If the question requests a specific form (e.g., 'common fraction in simplest radical form'), ensure the answer is in that form, unboxed, and without extra LaTeX formatting.
    - Do NOT enclose the final answer in boxes, \boxed{}, or extra math environments.
    - If multiple solutions exist, list all valid ones, separated by commas.
    - Always check for extraneous solutions and state which are valid.
    - Common pitfalls: boxing answers, failing to check domains, not simplifying, or including extra formatting.
    - At the end, clearly state the final answer(s) in the required form, and only include the answer value(s) in the 'answer' field.


2025/09/13 16:40:19 INFO dspy.evaluate.evaluate: Average Metric: 1.0 / 3 (33.3%)
GEPA Optimization:  22%|██▏       | 442/2000 [05:25<47:33,  1.83s/rollouts]

Iteration 8: New subsample score is not better, skipping
Iteration 9: Selected program 0 score: 0.66
Average Metric: 2.00 / 3 (66.7%): 100%|██████████| 3/3 [00:11<00:00,  3.76s/it] 

2025/09/13 16:40:30 INFO dspy.evaluate.evaluate: Average Metric: 2.0 / 3 (66.7%)



Iteration 9: Proposed new text for program: import dspy
from typing import Optional
from pydantic import constr

class MathQA(dspy.Signature):
    """
    Solve the given math problem step by step, showing all algebraic manipulations, substitutions, and logical deductions.
    - Begin by identifying relevant equations, identities, or proportional relationships.
    - Clearly define variables and state any assumptions.
    - Show all intermediate steps, especially when manipulating equations or solving for unknowns.
    - For problems involving fractions, always simplify the final answer to lowest terms.
    - For proportional reasoning, set up and solve the proportion explicitly.
    - Common pitfalls: arithmetic mistakes, algebraic sign errors, failing to exclude invalid solutions (e.g., extraneous roots, excluded values).
    - Edge cases: If a solution is not possible, state so explicitly.
    - Final answer must be clearly boxed or stated at the end, in the form of a simplified fr

2025/09/13 16:40:50 WARNING dspy.adapters.json_adapter: Failed to use structured output format, falling back to JSON mode.
2025/09/13 16:40:51 WARNING dspy.adapters.json_adapter: Failed to use structured output format, falling back to JSON mode.
2025/09/13 16:40:53 ERROR dspy.utils.parallelizer: Error for Example({'question': 'Adina and Lynn are going rock-climbing and need to rent special shoes. Lynn is normally size 9 and knows that her rock-climbing shoe-size is 42. If Adina is normally size 6, what size rock-climbing shoes should she rent assuming shoe size is directly proportional to rock-climbing shoe size?', 'reasoning': "Let $x$ be Adina's rock-climbing size.  The ratio of the girl's shoe sizes must be constant: \\[\\frac{\\text{Lynn's size}}{\\text{Adina's size}} = \\frac{9}{6}=\\frac{42}{x},\\]so $9x=42\\cdot 6$, or $x=\\frac{42\\cdot 6}{9}=\\boxed{28}$.", 'answer': '28'}) (input_keys={'question'}): litellm.BadRequestError: OpenAIException - Unrecognized request argument supp

Iteration 9: New subsample score is not better, skipping
Iteration 10: Selected program 0 score: 0.66
Average Metric: 2.00 / 3 (66.7%): 100%|██████████| 3/3 [00:08<00:00,  2.73s/it] 

2025/09/13 16:41:15 INFO dspy.evaluate.evaluate: Average Metric: 2.0 / 3 (66.7%)



Iteration 10: Proposed new text for program: import dspy
from typing import Optional
import re

class MathQA(dspy.Signature):
    """
    Solve the given math problem step by step, showing clear reasoning.
    - For function range/domain, always express the final answer in standard interval notation: use parentheses, no extra spaces, and no LaTeX formatting (e.g., (-infty,0)∪(0,infty)).
    - For numeric or algebraic answers, provide the simplest form.
    - Avoid LaTeX or extra formatting in the final answer field.
    - Common pitfalls: including LaTeX, extra spaces, or incorrect interval notation; skipping steps; not checking all cases.
    - Strategy: First, reason step by step. Then, extract the final answer in the required format.
    """
    question: str = dspy.InputField(desc="The math problem to solve")
    reasoning: str = dspy.OutputField(desc="Step-by-step solution and justification")
    answer: str = dspy.OutputField(desc="Final answer in required format (see instructio

2025/09/13 16:41:26 ERROR dspy.utils.parallelizer: Error for Example({'question': 'If $a\\star b = 9a+2b-ab+5$, what is the value of $5\\star1$?', 'reasoning': 'From the defined function, we know that $5\\star 1 = 9(5)+2(1)-(5)(1)+5= 45+2-5+5=\\boxed{47}$.', 'answer': '47'}) (input_keys={'question'}): bad escape \s at position 0
Traceback (most recent call last):
  File "/home/jacobjensen/Projects/prompt_evolution/gepa/.venv/lib/python3.11/site-packages/dspy/utils/parallelizer.py", line 55, in safe_func
    return user_function(item)
           ^^^^^^^^^^^^^^^^^^^
  File "/home/jacobjensen/Projects/prompt_evolution/gepa/.venv/lib/python3.11/site-packages/dspy/evaluate/evaluate.py", line 158, in process_item
    prediction = program(**example.inputs())
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/jacobjensen/Projects/prompt_evolution/gepa/.venv/lib/python3.11/site-packages/dspy/utils/callback.py", line 326, in sync_wrapper
    return fn(instance, *args, **kwargs)
         

Iteration 10: New subsample score is not better, skipping
Iteration 11: Selected program 1 score: 0.775
Average Metric: 3.00 / 3 (100.0%): 100%|██████████| 3/3 [00:02<00:00,  1.04it/s]

2025/09/13 16:41:37 INFO dspy.evaluate.evaluate: Average Metric: 3.0 / 3 (100.0%)
GEPA Optimization:  23%|██▎       | 457/2000 [06:43<1:13:46,  2.87s/rollouts]


Iteration 11: All subsample scores perfect. Skipping.
Iteration 11: Reflective mutation did not propose a new candidate
Iteration 12: Selected program 0 score: 0.66
Average Metric: 3.00 / 3 (100.0%): 100%|██████████| 3/3 [00:04<00:00,  1.39s/it]

2025/09/13 16:41:41 INFO dspy.evaluate.evaluate: Average Metric: 3.0 / 3 (100.0%)
GEPA Optimization:  23%|██▎       | 460/2000 [06:47<1:08:12,  2.66s/rollouts]


Iteration 12: All subsample scores perfect. Skipping.
Iteration 12: Reflective mutation did not propose a new candidate
Iteration 13: Selected program 1 score: 0.775
Average Metric: 2.00 / 3 (66.7%): 100%|██████████| 3/3 [00:03<00:00,  1.13s/it] 

2025/09/13 16:41:44 INFO dspy.evaluate.evaluate: Average Metric: 2.0 / 3 (66.7%)



Iteration 13: Proposed new text for program: import dspy
from typing import Literal

class MathQA(dspy.Signature):
    """
    Solve the given math word problem step by step, showing all necessary reasoning and calculations.
    Then, extract and return ONLY the final answer in the required format, with no extra text.

    Instructions:
    - Show clear, step-by-step reasoning in the 'reasoning' field.
    - In the 'answer' field, provide ONLY the final answer in the required format:
        * If the answer is a single number, output just the number (e.g., '42').
        * If the answer is an interval, output in standard interval notation with variable (e.g., 'x ∈ [-2,7]').
        * If the answer is a set, output in set notation (e.g., '{2, 3, 5}').
    - Do not include units, words, or extra formatting outside the required notation.
    - If there are multiple valid numeric answers, sum them and return the sum as a string.
    - Common pitfalls: Do not restate the question or includ

2025/09/13 16:42:09 INFO dspy.evaluate.evaluate: Average Metric: 1.0 / 3 (33.3%)
GEPA Optimization:  23%|██▎       | 466/2000 [07:15<1:22:35,  3.23s/rollouts]

Iteration 13: New subsample score is not better, skipping
Iteration 14: Selected program 1 score: 0.775
Average Metric: 3.00 / 3 (100.0%): 100%|██████████| 3/3 [00:06<00:00,  2.11s/it]

2025/09/13 16:42:15 INFO dspy.evaluate.evaluate: Average Metric: 3.0 / 3 (100.0%)
GEPA Optimization:  23%|██▎       | 469/2000 [07:21<1:17:33,  3.04s/rollouts]


Iteration 14: All subsample scores perfect. Skipping.
Iteration 14: Reflective mutation did not propose a new candidate
Iteration 15: Selected program 1 score: 0.775
Average Metric: 1.00 / 3 (33.3%): 100%|██████████| 3/3 [00:08<00:00,  2.90s/it]

2025/09/13 16:42:24 INFO dspy.evaluate.evaluate: Average Metric: 1.0 / 3 (33.3%)



Iteration 15: Proposed new text for program: import dspy
import re
from typing import Optional

class MathQA(dspy.Signature):
    """
    Solve the given math word problem step by step, showing all necessary reasoning and calculations.
    Then, extract and return ONLY the final numeric answer as a string, with no units or extra text.

    Instructions:
    - Show clear, step-by-step reasoning in the 'reasoning' field.
    - In the 'answer' field, provide ONLY the final numeric answer (e.g., '42', '6', '32', 'sqrt{10}'), not a sentence or explanation.
    - For answers involving radicals, use the format 'sqrt{n}' (e.g., 'sqrt{10}'), not LaTeX or parentheses.
    - For area change, always report the magnitude (positive value) unless the problem explicitly asks for increase/decrease or sign.
    - Do not include units, words, or extra formatting in the answer.
    - If there are multiple valid numeric answers, sum them and return the sum as a string.
    - Common pitfalls: Do not restat

2025/09/13 16:42:50 INFO dspy.evaluate.evaluate: Average Metric: 2.0 / 3 (66.7%)
2025/09/13 16:43:22 INFO dspy.evaluate.evaluate: Average Metric: 120.0 / 200 (60.0%)
GEPA Optimization:  34%|███▍      | 675/2000 [08:28<10:31,  2.10rollouts/s]  

Iteration 15: Full valset score for new program: 0.6
Iteration 15: Full train_val score for new program: 0.6
Iteration 15: Individual valset scores for new program: [True, False, True, True, False, False, False, True, True, True, False, False, True, False, False, False, False, True, False, True, False, True, True, True, True, True, True, False, True, True, True, False, False, True, True, True, True, True, False, False, True, False, True, True, False, True, False, True, True, True, False, False, True, True, True, True, True, True, False, True, True, False, False, True, False, False, False, False, True, False, True, False, True, True, True, True, True, False, True, False, False, True, True, False, True, True, True, True, True, False, False, True, True, True, True, True, True, False, False, True, False, False, True, True, True, True, True, False, True, False, True, False, False, True, False, False, True, True, True, True, True, False, False, True, True, True, True, True, False, True, True

2025/09/13 16:43:25 INFO dspy.evaluate.evaluate: Average Metric: 3.0 / 3 (100.0%)
GEPA Optimization:  34%|███▍      | 678/2000 [08:30<10:41,  2.06rollouts/s]


Iteration 16: All subsample scores perfect. Skipping.
Iteration 16: Reflective mutation did not propose a new candidate
Iteration 17: Selected program 0 score: 0.66
Average Metric: 3.00 / 3 (100.0%): 100%|██████████| 3/3 [00:06<00:00,  2.06s/it]

2025/09/13 16:43:31 INFO dspy.evaluate.evaluate: Average Metric: 3.0 / 3 (100.0%)
GEPA Optimization:  34%|███▍      | 681/2000 [08:37<11:35,  1.90rollouts/s]


Iteration 17: All subsample scores perfect. Skipping.
Iteration 17: Reflective mutation did not propose a new candidate
Iteration 18: Selected program 0 score: 0.66
Average Metric: 2.00 / 3 (66.7%): 100%|██████████| 3/3 [00:01<00:00,  2.06it/s] 

2025/09/13 16:43:32 INFO dspy.evaluate.evaluate: Average Metric: 2.0 / 3 (66.7%)



Iteration 18: Proposed new text for program: import dspy
from typing import Optional

class QuantitativeReasoningSignature(dspy.Signature):
    """
    Solve quantitative math word problems step by step.
    - Carefully extract all relevant quantities and their time references from the question.
    - Pay special attention to whether the current value is the original or already increased.
    - For problems involving repeated operations (e.g., doubling), determine the correct starting value and number of steps.
    - Avoid common pitfalls: do not double the original value if the current value is already doubled; always use the most recent value as the base for future operations.
    - Show all intermediate calculations and reasoning.
    - Provide a concise, boxed final answer in dollars if applicable.
    """
    question: str = dspy.InputField(desc="A math word problem requiring stepwise quantitative reasoning.")
    computed_answer: str = dspy.InputField(desc="The final numeric ans

2025/09/13 16:43:44 INFO dspy.evaluate.evaluate: Average Metric: 1.0 / 3 (33.3%)
GEPA Optimization:  34%|███▍      | 687/2000 [08:50<14:10,  1.54rollouts/s]

Iteration 18: New subsample score is not better, skipping
Iteration 19: Selected program 1 score: 0.775
Average Metric: 2.00 / 3 (66.7%): 100%|██████████| 3/3 [00:07<00:00,  2.66s/it]

2025/09/13 16:43:52 INFO dspy.evaluate.evaluate: Average Metric: 2.0 / 3 (66.7%)



Iteration 19: Proposed new text for program: import dspy
import re
from typing import Optional

class MathQA(dspy.Signature):
    """
    Solve the given math word problem step by step, showing all necessary reasoning and calculations.
    Then, extract and return ONLY the final answer in the required format (number, variable, coordinate, or list), with no units or extra text.

    Instructions:
    - Show clear, step-by-step reasoning in the 'reasoning' field.
    - In the 'answer' field, provide ONLY the final answer in the format required by the question:
        - If the answer is a number, return just the number (e.g., '42', '6', '0').
        - If the answer is a coordinate, return it in the form '(x,y)' (e.g., '(9,11)').
        - If the answer is a variable or object, return just its name (e.g., 'x', 'A').
        - If the answer is a list or set, return the elements separated by commas, or as a sum if specified.
    - Do not include units, words, or extra formatting.
    - Do

2025/09/13 16:44:18 INFO dspy.evaluate.evaluate: Average Metric: 3.0 / 3 (100.0%)
2025/09/13 16:45:00 INFO dspy.evaluate.evaluate: Average Metric: 178.0 / 200 (89.0%)
GEPA Optimization:  45%|████▍     | 893/2000 [10:06<07:56,  2.32rollouts/s]

Iteration 19: New program is on the linear pareto front
Iteration 19: Full valset score for new program: 0.89
Iteration 19: Full train_val score for new program: 0.89
Iteration 19: Individual valset scores for new program: [True, False, True, True, False, False, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, False, True, True, True, True, True, True, False, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, False, True, True, True, True, True, True, True, False, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, False, True, True, True, True, True, True, True, True, True, True, False, True, True, True, True, True, True, True, False, False, True, True, True, True, False, True, True, True, True, True, True, True, True, True, True, True, Tr

2025/09/13 16:45:04 INFO dspy.evaluate.evaluate: Average Metric: 2.0 / 3 (66.7%)



Iteration 20: Proposed new text for program: import dspy
import re
from typing import Optional

class MathQA(dspy.Signature):
    """
    Solve the given math word problem step by step, showing all necessary reasoning and calculations.
    Then, extract and return ONLY the final answer in the required format (number, variable, coordinate, list, fraction, or radical), with no units or extra text.

    Instructions:
    - Show clear, step-by-step reasoning in the 'reasoning' field.
    - In the 'answer' field, provide ONLY the final answer in the format required by the question:
        - If the answer is a number, return just the number (e.g., '42', '6', '0').
        - If the answer is a coordinate, return it in the form '(x,y)' (e.g., '(9,11)').
        - If the answer is a variable or object, return just its name (e.g., 'x', 'A').
        - If the answer is a list or set, return the elements separated by commas, or as a sum if specified.
        - If the answer is a fraction, use 'a

2025/09/13 16:45:41 INFO dspy.evaluate.evaluate: Average Metric: 1.0 / 3 (33.3%)
GEPA Optimization:  45%|████▍     | 899/2000 [10:47<11:37,  1.58rollouts/s]

Iteration 20: New subsample score is not better, skipping
Iteration 21: Selected program 0 score: 0.66
Average Metric: 3.00 / 3 (100.0%): 100%|██████████| 3/3 [00:02<00:00,  1.07it/s]

2025/09/13 16:45:44 INFO dspy.evaluate.evaluate: Average Metric: 3.0 / 3 (100.0%)
GEPA Optimization:  45%|████▌     | 902/2000 [10:50<11:43,  1.56rollouts/s]


Iteration 21: All subsample scores perfect. Skipping.
Iteration 21: Reflective mutation did not propose a new candidate
Iteration 22: Selected program 3 score: 0.89
Average Metric: 3.00 / 3 (100.0%): 100%|██████████| 3/3 [00:04<00:00,  1.59s/it]

2025/09/13 16:45:49 INFO dspy.evaluate.evaluate: Average Metric: 3.0 / 3 (100.0%)
GEPA Optimization:  45%|████▌     | 905/2000 [10:55<12:13,  1.49rollouts/s]


Iteration 22: All subsample scores perfect. Skipping.
Iteration 22: Reflective mutation did not propose a new candidate
Iteration 23: Selected program 0 score: 0.66
Average Metric: 1.00 / 3 (33.3%): 100%|██████████| 3/3 [00:05<00:00,  1.69s/it] 

2025/09/13 16:45:54 INFO dspy.evaluate.evaluate: Average Metric: 1.0 / 3 (33.3%)



Iteration 23: Proposed new text for program: import dspy
from typing import Literal

class MathQA(dspy.Signature):
    """
    Solve the given math problem step by step, showing all intermediate steps and calculations.
    - For equations, show algebraic manipulations and check for extraneous solutions.
    - For inequalities, solve each separately, reduce all fractions, and combine using interval notation.
    - For word problems, write out the formula, substitute values, and show all calculations.
    - Always reduce fractions to lowest terms.
    - For interval notation, use parentheses for open intervals and brackets for closed intervals, with no extra spaces.
    - For numeric answers, provide a plain number (no LaTeX, no words, no dollar sign, no box).
    - For interval answers, use the form (a,b], [a,b), etc., with reduced fractions and no LaTeX.
    - For rounding, follow the instructions exactly (e.g., "nearest dollar").
    - Do not include boxed answers, LaTeX, or explanat

2025/09/13 16:46:14 WARNING dspy.adapters.json_adapter: Failed to use structured output format, falling back to JSON mode.
2025/09/13 16:46:14 WARNING dspy.adapters.json_adapter: Failed to use structured output format, falling back to JSON mode.
2025/09/13 16:46:15 WARNING dspy.adapters.json_adapter: Failed to use structured output format, falling back to JSON mode.
2025/09/13 16:46:18 ERROR dspy.utils.parallelizer: Error for Example({'question': "Alan saved 500 dollars in a bank account that compounds 3 percent annually.  Assuming there are no other transactions, after 10 years, how much is in Alan's bank account?  (Give your answer to the nearest dollar.)", 'reasoning': 'After ten years, at a three percent annual interest rate, the bank account will have grown to $500 \\cdot 1.03^{10} = \\boxed{672}$, to the nearest dollar.', 'answer': '672'}) (input_keys={'question'}): litellm.BadRequestError: OpenAIException - Unrecognized request argument supplied: instructions
Traceback (most rec

Iteration 23: New subsample score is not better, skipping
Iteration 24: Selected program 1 score: 0.775
Average Metric: 2.00 / 3 (66.7%): 100%|██████████| 3/3 [00:07<00:00,  2.54s/it] 

2025/09/13 16:46:26 INFO dspy.evaluate.evaluate: Average Metric: 2.0 / 3 (66.7%)



Iteration 24: Proposed new text for program: import dspy
from typing import Optional

class MathQA(dspy.Signature):
    """
    Solve the given math word problem step by step, showing all necessary reasoning and calculations.
    Then, extract and return ONLY the final answer as a string, matching the exact form requested in the question (e.g., as a common fraction, in terms of radicals, etc.), with no units or extra text.

    Instructions:
    - Show clear, step-by-step reasoning in the 'reasoning' field.
    - In the 'answer' field, provide ONLY the final answer as a string, matching the format requested in the question (e.g., if the question says "as a common fraction", answer "25/8"; if "in terms of radicals", answer "2 + sqrt(3)"; if boxed or LaTeX is requested, use that format).
    - Do NOT convert fractions to decimals unless explicitly requested.
    - Do NOT include units, words, or extra formatting (no sentences, no "The answer is", no boxing, unless the question requests 

2025/09/13 16:46:56 INFO dspy.evaluate.evaluate: Average Metric: 3.0 / 3 (100.0%)
2025/09/13 16:49:04 INFO dspy.evaluate.evaluate: Average Metric: 165.0 / 200 (82.5%)
GEPA Optimization:  56%|█████▌    | 1117/2000 [14:10<12:26,  1.18rollouts/s]

Iteration 24: Full valset score for new program: 0.825
Iteration 24: Full train_val score for new program: 0.825
Iteration 24: Individual valset scores for new program: [True, True, True, True, False, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, False, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, False, True, False, True, False, False, True, True, True, True, True, True, False, True, True, True, True, True, False, True, True, True, True, True, False, True, True, False, False, True, True, False, True, False, False, True, True, True, True, True, True, False, True, True, True, True, True, True, True, False, True, True, False, True, False, False, True, True, True, False, False, True, True, True, True, True, False, True, True, True, True, True, False, True, True, True, True, True, True, True, True, True, True, True, True, Tru

2025/09/13 16:49:07 INFO dspy.evaluate.evaluate: Average Metric: 2.0 / 3 (66.7%)



Iteration 25: Proposed new text for program: import dspy
from typing import Optional

class MathQA(dspy.Signature):
    """
    Solve the given math problem step by step, showing all necessary reasoning and calculations.
    - Carefully analyze the problem and identify all relevant constraints and requirements.
    - For domain/range questions, express the answer as a union of intervals in standard mathematical notation, without LaTeX or boxed formatting unless explicitly requested.
    - For value questions, provide the final answer as a plain number or expression, not wrapped in LaTeX or boxes.
    - Avoid including extra formatting such as \boxed{} or \[ \] unless the question specifically asks for it.
    - Common pitfalls: hallucinating boxed/LaTeX answers, skipping steps, or misinterpreting the required answer format.
    - Edge cases: If multiple answers are possible, list all; if the answer is an interval, use standard interval notation.
    - Successful strategy: First, write

2025/09/13 16:49:24 WARNING dspy.adapters.json_adapter: Failed to use structured output format, falling back to JSON mode.
2025/09/13 16:49:25 WARNING dspy.adapters.json_adapter: Failed to use structured output format, falling back to JSON mode.
2025/09/13 16:49:26 WARNING dspy.adapters.json_adapter: Failed to use structured output format, falling back to JSON mode.
2025/09/13 16:49:28 ERROR dspy.utils.parallelizer: Error for Example({'question': 'If $F(a, b, c, d) = a^b + c ^ d$, what is the value of $b$ such that $F(4, b, 2, 3) = 12$?', 'reasoning': 'Plugging in, we have that $4^b + 2^3 = 12$.  This rearranges to $4^b = 4$, or $b = \\boxed{1}$.', 'answer': '1'}) (input_keys={'question'}): litellm.BadRequestError: OpenAIException - Unrecognized request argument supplied: instructions
Traceback (most recent call last):
  File "/home/jacobjensen/Projects/prompt_evolution/gepa/.venv/lib/python3.11/site-packages/litellm/llms/openai/openai.py", line 736, in completion
    raise e
  File "/

Iteration 25: New subsample score is not better, skipping
Iteration 26: Selected program 4 score: 0.825
Average Metric: 3.00 / 3 (100.0%): 100%|██████████| 3/3 [00:03<00:00,  1.05s/it]

2025/09/13 16:49:33 INFO dspy.evaluate.evaluate: Average Metric: 3.0 / 3 (100.0%)
GEPA Optimization:  56%|█████▋    | 1126/2000 [14:39<13:55,  1.05rollouts/s]


Iteration 26: All subsample scores perfect. Skipping.
Iteration 26: Reflective mutation did not propose a new candidate
Iteration 27: Selected program 3 score: 0.89
Average Metric: 1.00 / 3 (33.3%): 100%|██████████| 3/3 [00:05<00:00,  1.84s/it]

2025/09/13 16:49:38 INFO dspy.evaluate.evaluate: Average Metric: 1.0 / 3 (33.3%)



Iteration 27: Proposed new text for program: import dspy
import re
from typing import Optional, Literal

class MathQA(dspy.Signature):
    """
    Solve the given math word problem step by step, showing all necessary reasoning and calculations.
    Then, extract and return ONLY the final answer in the required format, as specified in the 'answer_format' field.

    Instructions:
    - Show clear, step-by-step reasoning in the 'reasoning' field.
    - In the 'answer' field, provide ONLY the final answer in the format required by 'answer_format':
        - If 'answer_format' is 'integer', return just the integer (e.g., '42').
        - If 'answer_format' is 'common fraction', return a reduced fraction (e.g., '3/2').
        - If 'answer_format' is 'coordinate', return in the form '(x,y)' (e.g., '(9,11)').
        - If 'answer_format' is 'variable', return just the variable name (e.g., 'x').
        - If 'answer_format' is 'list', return elements separated by commas (e.g., '2,3,5').
    

2025/09/13 16:50:12 INFO dspy.evaluate.evaluate: Average Metric: 2.0 / 3 (66.7%)
2025/09/13 16:52:32 INFO dspy.evaluate.evaluate: Average Metric: 133.0 / 200 (66.5%)
GEPA Optimization:  67%|██████▋   | 1332/2000 [17:38<10:00,  1.11rollouts/s]

Iteration 27: Full valset score for new program: 0.665
Iteration 27: Full train_val score for new program: 0.665
Iteration 27: Individual valset scores for new program: [True, True, True, False, False, False, False, True, True, True, True, False, True, True, False, True, False, True, True, True, False, True, True, True, True, True, True, True, True, True, True, False, True, True, True, True, True, True, True, True, True, True, True, True, True, True, False, True, True, True, False, False, False, False, True, True, True, False, False, True, True, False, True, True, False, True, False, True, True, False, True, False, False, True, False, True, True, False, True, False, False, True, True, True, True, False, False, False, True, False, False, True, True, True, True, True, True, True, False, True, False, False, True, True, False, True, False, False, True, False, True, False, True, True, False, False, True, True, True, False, True, True, True, True, True, False, True, True, True, True, False, 

2025/09/13 16:52:36 INFO dspy.evaluate.evaluate: Average Metric: 3.0 / 3 (100.0%)
GEPA Optimization:  67%|██████▋   | 1335/2000 [17:42<10:01,  1.11rollouts/s]


Iteration 28: All subsample scores perfect. Skipping.
Iteration 28: Reflective mutation did not propose a new candidate
Iteration 29: Selected program 0 score: 0.66
Average Metric: 3.00 / 3 (100.0%): 100%|██████████| 3/3 [00:04<00:00,  1.34s/it]

2025/09/13 16:52:40 INFO dspy.evaluate.evaluate: Average Metric: 3.0 / 3 (100.0%)
GEPA Optimization:  67%|██████▋   | 1338/2000 [17:46<10:04,  1.10rollouts/s]


Iteration 29: All subsample scores perfect. Skipping.
Iteration 29: Reflective mutation did not propose a new candidate
Iteration 30: Selected program 3 score: 0.89
Average Metric: 2.00 / 2 (100.0%):  67%|██████▋   | 2/3 [00:03<00:01,  1.61s/it]

2025/09/13 16:52:47 WARNING dspy.clients.lm: LM response was truncated due to exceeding max_tokens=32000. You can inspect the latest LM interactions with `dspy.inspect_history()`. To avoid truncation, consider passing a larger max_tokens when setting up dspy.LM. You may also consider increasing the temperature (currently 0.0)  if the reason for truncation is repetition.


Average Metric: 2.00 / 3 (66.7%): 100%|██████████| 3/3 [00:27<00:00,  9.03s/it] 

2025/09/13 16:53:07 INFO dspy.evaluate.evaluate: Average Metric: 2.0 / 3 (66.7%)



Iteration 30: Proposed new text for program: import dspy
import re
from typing import List, Tuple, Optional

class ExtractEndpoints(dspy.Signature):
    """
    Given a math word problem describing a piecewise linear function (e.g., a graph with line segments), extract the list of all endpoints as (x, y) coordinate pairs, in order.

    Instructions:
    - Return the endpoints as a list of coordinate pairs, in the order they appear in the problem.
    - Each coordinate should be in the form (x, y), with integer values.
    - Do not include any extra text, explanations, or formatting.
    - Example: "The points are (-5,-4), (-2,5), (-1,3), (1,-5), (3,2), (5,2)." → [(-5,-4), (-2,5), (-1,3), (1,-5), (3,2), (5,2)]
    """
    question: str = dspy.InputField(desc="The math word problem with the graph or endpoints")
    endpoints: str = dspy.OutputField(desc="List of endpoints as (x,y) pairs, in order, no extra text")

class MathQARefined(dspy.Signature):
    """
    Solve the given math wo

2025/09/13 16:53:48 INFO dspy.evaluate.evaluate: Average Metric: 2.0 / 3 (66.7%)
GEPA Optimization:  67%|██████▋   | 1344/2000 [18:54<16:00,  1.46s/rollouts]

Iteration 30: New subsample score is not better, skipping
Iteration 31: Selected program 3 score: 0.89
Average Metric: 3.00 / 3 (100.0%): 100%|██████████| 3/3 [00:03<00:00,  1.08s/it]

2025/09/13 16:53:51 INFO dspy.evaluate.evaluate: Average Metric: 3.0 / 3 (100.0%)
GEPA Optimization:  67%|██████▋   | 1347/2000 [18:57<15:47,  1.45s/rollouts]


Iteration 31: All subsample scores perfect. Skipping.
Iteration 31: Reflective mutation did not propose a new candidate
Iteration 32: Selected program 4 score: 0.825
Average Metric: 2.00 / 3 (66.7%): 100%|██████████| 3/3 [00:08<00:00,  2.76s/it] 

2025/09/13 16:53:59 INFO dspy.evaluate.evaluate: Average Metric: 2.0 / 3 (66.7%)



Iteration 32: Proposed new text for program: import dspy
from typing import Optional

class MathQAReasoning(dspy.Signature):
    """
    Solve the given math word problem step by step, showing all necessary reasoning and calculations.
    Then, extract and return ONLY the final answer as a string, matching the exact form requested in the question (e.g., as a common fraction, in terms of radicals, LaTeX, or boxed, etc.), with no units or extra text.

    Instructions:
    - Show clear, step-by-step reasoning in the 'reasoning' field.
    - In the 'raw_answer' field, provide ONLY the final answer as a string, matching the format requested in the question (e.g., if the question says "as a common fraction", answer "25/8"; if "in terms of radicals", answer "2 + sqrt(3)"; if boxed or LaTeX is requested, use that format).
    - Do NOT convert fractions to decimals unless explicitly requested.
    - Do NOT include units, words, or extra formatting (no sentences, no "The answer is", no boxing,

2025/09/13 16:54:33 INFO dspy.evaluate.evaluate: Average Metric: 2.0 / 3 (66.7%)
GEPA Optimization:  68%|██████▊   | 1353/2000 [19:39<21:21,  1.98s/rollouts]

Iteration 32: New subsample score is not better, skipping
Iteration 33: Selected program 3 score: 0.89
Average Metric: 1.00 / 3 (33.3%): 100%|██████████| 3/3 [00:07<00:00,  2.66s/it] 

2025/09/13 16:54:41 INFO dspy.evaluate.evaluate: Average Metric: 1.0 / 3 (33.3%)



Iteration 33: Proposed new text for program: import dspy
import re
from typing import Optional

class MathQA(dspy.Signature):
    """
    Solve the given math word problem step by step, showing all necessary reasoning and calculations.
    Then, extract and return ONLY the final answer in the required format (number, variable, coordinate, interval, or list), with no units or extra text.

    Instructions:
    - Show clear, step-by-step reasoning in the 'reasoning' field.
    - In the 'answer' field, provide ONLY the final answer in the format required by the question:
        - If the answer is a number, return just the number (e.g., '42', '6', '0').
        - If the answer is a coordinate, return it in the form '(x,y)' (e.g., '(9,11)').
        - If the answer is a variable or object, return just its name (e.g., 'x', 'A').
        - If the answer is a list or set, return the elements separated by spaces (e.g., '2 3 5'), unless the question specifies a sum.
        - If the answer is 

2025/09/13 16:55:21 INFO dspy.evaluate.evaluate: Average Metric: 2.0 / 3 (66.7%)
2025/09/13 16:55:24 WARNING dspy.clients.lm: LM response was truncated due to exceeding max_tokens=32000. You can inspect the latest LM interactions with `dspy.inspect_history()`. To avoid truncation, consider passing a larger max_tokens when setting up dspy.LM. You may also consider increasing the temperature (currently 0.0)  if the reason for truncation is repetition.
2025/09/13 16:55:46 INFO dspy.evaluate.evaluate: Average Metric: 179.0 / 200 (89.5%)
GEPA Optimization:  78%|███████▊  | 1559/2000 [20:51<04:41,  1.57rollouts/s]

Iteration 33: New program is on the linear pareto front
Iteration 33: Full valset score for new program: 0.895
Iteration 33: Full train_val score for new program: 0.895
Iteration 33: Individual valset scores for new program: [True, True, True, False, False, True, True, True, True, True, True, True, True, False, False, True, True, True, True, True, False, True, True, True, True, True, True, True, True, True, True, False, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, False, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, False, True, True, False, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, False, False, False, True, True, True, False, True, True, True, True, True, True, True, True, True, True, True,

2025/09/13 16:55:51 INFO dspy.evaluate.evaluate: Average Metric: 3.0 / 3 (100.0%)
GEPA Optimization:  78%|███████▊  | 1562/2000 [20:57<04:48,  1.52rollouts/s]


Iteration 34: All subsample scores perfect. Skipping.
Iteration 34: Reflective mutation did not propose a new candidate
Iteration 35: Selected program 4 score: 0.825
Average Metric: 3.00 / 3 (100.0%): 100%|██████████| 3/3 [00:02<00:00,  1.03it/s]

2025/09/13 16:55:54 INFO dspy.evaluate.evaluate: Average Metric: 3.0 / 3 (100.0%)
GEPA Optimization:  78%|███████▊  | 1565/2000 [21:00<04:49,  1.50rollouts/s]


Iteration 35: All subsample scores perfect. Skipping.
Iteration 35: Reflective mutation did not propose a new candidate
Iteration 36: Selected program 0 score: 0.66
Average Metric: 2.00 / 3 (66.7%): 100%|██████████| 3/3 [00:08<00:00,  2.98s/it] 

2025/09/13 16:56:03 INFO dspy.evaluate.evaluate: Average Metric: 2.0 / 3 (66.7%)



Iteration 36: Proposed new text for program: import dspy
import re
from typing import Literal

class MathQA(dspy.Signature):
    """
    Solve the given math problem step by step, showing clear reasoning and calculations.
    - Provide detailed reasoning, breaking down the problem into logical steps.
    - For the final answer, extract and present it as a plain text value (e.g., '1023/1024', '7', '50'), not in LaTeX or math mode.
    - Avoid enclosing the answer in any LaTeX delimiters (e.g., $, \(\), \[\]).
    - If the answer is a fraction, write it as 'numerator/denominator' (e.g., '1023/1024').
    - Common pitfalls: Do not include units unless specified; do not repeat the question; do not use LaTeX in the answer field.
    - Edge cases: If the answer is an integer, output it as a plain integer string.
    - Successful strategy: Reason step by step, then clearly state the answer in the required format.
    """
    question: str = dspy.InputField(desc="The math problem to solve")
 

2025/09/13 16:56:15 INFO dspy.evaluate.evaluate: Average Metric: 3.0 / 3 (100.0%)
2025/09/13 16:56:38 INFO dspy.evaluate.evaluate: Average Metric: 132.0 / 200 (66.0%)
GEPA Optimization:  89%|████████▊ | 1771/2000 [21:44<01:20,  2.86rollouts/s]

Iteration 36: Full valset score for new program: 0.66
Iteration 36: Full train_val score for new program: 0.66
Iteration 36: Individual valset scores for new program: [True, True, True, True, False, False, False, True, True, True, False, False, True, True, False, True, True, True, False, True, False, True, True, True, False, True, True, False, True, True, True, False, False, True, True, True, True, True, False, True, True, False, True, True, False, True, False, True, True, True, False, False, True, True, True, True, True, True, False, True, True, False, False, True, False, False, False, False, True, False, True, False, True, True, True, True, False, False, True, False, False, True, True, True, True, True, True, True, True, False, False, True, True, True, True, True, True, False, False, True, False, True, True, True, True, True, True, False, True, False, False, False, True, True, False, False, True, True, True, True, True, False, False, True, True, True, True, False, True, True, True, T

2025/09/13 16:56:56 INFO dspy.evaluate.evaluate: Average Metric: 2.0 / 3 (66.7%)



Iteration 37: Proposed new text for program: import dspy
from typing import Optional

class MathQAReasoning(dspy.Signature):
    """
    Solve the given math word problem step by step, showing all necessary reasoning and calculations.
    - Provide a detailed, logically ordered, step-by-step solution in the 'reasoning' field.
    - Include all algebraic manipulations, casework, and justification for each step.
    - For function inverses or piecewise functions, be explicit about domain, sign, and order of terms.
    - Do not state the final answer in the reasoning; reserve it for the answer extraction step.
    - Avoid skipping steps, and clarify any sign or domain choices.
    """
    question: str = dspy.InputField(desc="The math word problem to solve")
    reasoning: str = dspy.OutputField(desc="Step-by-step solution and calculations, ending just before the final answer")

class MathQAAnswer(dspy.Signature):
    """
    Given the question and the full step-by-step reasoning, extrac

2025/09/13 16:57:27 INFO dspy.evaluate.evaluate: Average Metric: 2.0 / 3 (66.7%)
GEPA Optimization:  89%|████████▉ | 1777/2000 [22:33<02:07,  1.75rollouts/s]

Iteration 37: New subsample score is not better, skipping
Iteration 38: Selected program 6 score: 0.895
Average Metric: 2.00 / 3 (66.7%): 100%|██████████| 3/3 [00:16<00:00,  5.43s/it] 

2025/09/13 16:57:44 INFO dspy.evaluate.evaluate: Average Metric: 2.0 / 3 (66.7%)



Iteration 38: Proposed new text for program: import dspy
import re
from typing import Optional, Tuple

class MathQA(dspy.Signature):
    """
    Solve the given math word problem step by step, showing all necessary reasoning and calculations.
    Then, extract and return ONLY the final answer in the required format (number, variable, coordinate, interval, list, or algebraic expression), with no units or extra text.

    Instructions:
    - Show clear, step-by-step reasoning in the 'reasoning' field.
    - In the 'answer' field, provide ONLY the final answer in the format required by the question:
        - If the answer is a number, return just the number (e.g., '42', '6', '0').
        - If the answer is a coordinate, return it in the form '(x,y)' (e.g., '(9,11)').
        - If the answer is a variable or object, return just its name (e.g., 'x', 'A').
        - If the answer is a list or set, return the elements separated by spaces (e.g., '2 3 5'), unless the question specifies a sum

2025/09/13 16:59:48 INFO dspy.evaluate.evaluate: Average Metric: 1.0 / 3 (33.3%)
GEPA Optimization:  89%|████████▉ | 1783/2000 [24:54<05:16,  1.46s/rollouts]

Iteration 38: New subsample score is not better, skipping
Iteration 39: Selected program 7 score: 0.66
Average Metric: 1.00 / 3 (33.3%): 100%|██████████| 3/3 [00:08<00:00,  2.70s/it] 

2025/09/13 16:59:56 INFO dspy.evaluate.evaluate: Average Metric: 1.0 / 3 (33.3%)



Iteration 39: Proposed new text for program: import dspy
import re
from typing import Optional

class MathQA(dspy.Signature):
    """
    Solve the given math problem step by step, showing clear reasoning and calculations.
    - Provide detailed reasoning, breaking down the problem into logical steps.
    - For the final answer, strictly follow the required answer_format (see below).
    - Do NOT include LaTeX delimiters ($, \\(\\), \\[\\]), units, or extra text in the answer.
    - If the answer is a fraction, write as 'numerator/denominator' (e.g., '1023/1024').
    - If the answer is an integer, output as a plain integer string (e.g., '7').
    - If the answer is an ordered pair, output as '(a,b)' with no spaces (e.g., '(1,1)').
    - If the answer is a factorization, output as '(ax^2+bx+c)(dx^2+ex+f)' with integer coefficients and no extra text.
    - Common pitfalls: Do not repeat the question; do not use LaTeX in the answer; do not add explanations in the answer field.
    - Edg

2025/09/13 17:00:28 INFO dspy.evaluate.evaluate: Average Metric: 3.0 / 3 (100.0%)
2025/09/13 17:01:09 INFO dspy.evaluate.evaluate: Average Metric: 150.0 / 200 (75.0%)
GEPA Optimization:  99%|█████████▉| 1989/2000 [26:15<00:08,  1.31rollouts/s]

Iteration 39: Full valset score for new program: 0.75
Iteration 39: Full train_val score for new program: 0.75
Iteration 39: Individual valset scores for new program: [True, True, True, False, False, False, False, True, True, True, True, False, True, True, False, False, False, True, True, True, False, True, True, True, False, True, True, True, True, True, True, False, True, True, True, True, True, True, True, True, True, True, True, True, True, True, False, True, True, True, True, True, False, False, True, True, True, True, True, True, True, False, True, False, True, True, False, True, True, True, True, True, True, True, True, True, True, False, True, False, False, True, True, True, True, True, False, True, True, False, False, True, True, True, True, True, True, True, False, True, True, False, True, True, False, True, True, False, True, False, True, False, True, True, False, False, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, False, Tr

2025/09/13 17:01:48 INFO dspy.evaluate.evaluate: Average Metric: 2.0 / 3 (66.7%)



Iteration 40: Proposed new text for program: import dspy
import re
from typing import Optional, Dict, Any

class MathQA(dspy.Signature):
    """
    Solve the given math word problem step by step, showing all necessary reasoning and calculations.
    Then, extract and return ONLY the final answer in the required format (number, variable, coordinate, interval, or list), with no units or extra text.

    Instructions:
    - Show clear, step-by-step reasoning in the 'reasoning' field.
    - In the 'answer' field, provide ONLY the final answer in the format required by the question:
        - If the answer is a number, return just the number (e.g., '42', '6', '0').
        - If the answer is a coordinate, return it in the form '(x,y)' (e.g., '(9,11)').
        - If the answer is a variable or object, return just its name (e.g., 'x', 'A').
        - If the answer is a list or set, return the elements separated by spaces (e.g., '2 3 5'), unless the question specifies a sum.
        - If the

2025/09/13 17:02:49 INFO dspy.evaluate.evaluate: Average Metric: 2.0 / 3 (66.7%)
GEPA Optimization: 100%|█████████▉| 1995/2000 [27:54<00:05,  1.18s/rollouts]

Iteration 40: New subsample score is not better, skipping
Iteration 41: Selected program 6 score: 0.895
Average Metric: 3.00 / 3 (100.0%): 100%|██████████| 3/3 [00:07<00:00,  2.35s/it]

2025/09/13 17:02:56 INFO dspy.evaluate.evaluate: Average Metric: 3.0 / 3 (100.0%)
GEPA Optimization: 100%|█████████▉| 1998/2000 [28:01<00:02,  1.20s/rollouts]


Iteration 41: All subsample scores perfect. Skipping.
Iteration 41: Reflective mutation did not propose a new candidate
Iteration 42: Selected program 4 score: 0.825
Average Metric: 1.00 / 3 (33.3%): 100%|██████████| 3/3 [00:08<00:00,  2.83s/it]

2025/09/13 17:03:04 INFO dspy.evaluate.evaluate: Average Metric: 1.0 / 3 (33.3%)



Iteration 42: Proposed new text for program: import dspy
from typing import Optional

class MathQAReasoning(dspy.Signature):
    """
    Solve the given math word problem step by step, showing all necessary reasoning and calculations.
    Do NOT provide the final answer yet; focus only on the step-by-step solution and intermediate results.

    Instructions:
    - Show clear, step-by-step reasoning in the 'reasoning' field.
    - Do NOT include the final answer or summary sentence.
    - Use mathematical notation (e.g., fractions, radicals, $\pi$, $\LaTeX$) as appropriate.
    - Avoid restating the question.
    - Example:
        Question: "What is the value of $\\frac{1}{2} + \\frac{1}{3}$? Express your answer as a common fraction."
        Reasoning: "Find common denominator: 1/2 + 1/3 = 3/6 + 2/6 = 5/6."
    """
    question: str = dspy.InputField(desc="The math word problem to solve")
    reasoning: str = dspy.OutputField(desc="Step-by-step solution and calculations, no final ans

2025/09/13 17:03:30 INFO dspy.evaluate.evaluate: Average Metric: 3.0 / 3 (100.0%)
2025/09/13 17:04:13 INFO dspy.evaluate.evaluate: Average Metric: 176.0 / 200 (88.0%)
GEPA Optimization: 100%|█████████▉| 1998/2000 [29:18<00:01,  1.14rollouts/s]

Iteration 42: Full valset score for new program: 0.88
Iteration 42: Full train_val score for new program: 0.88
Iteration 42: Individual valset scores for new program: [True, False, True, True, False, True, True, True, False, True, True, True, True, True, True, False, False, True, True, True, True, True, True, True, False, True, True, True, True, True, True, True, True, True, True, True, True, False, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, False, True, True, True, True, True, True, True, False, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, False, True, True, False, True, False, True, True, True, True, True, True, True, False, True, True, True, True, True, True, True, True, True, True, False, True, True, True, True, True, True, True, False, False, True, True, True, True, False, True, False, True, True, True, True, True, True, True, True, True, False, True, True, True, True, True, True, True, Tru

Let's see the DSPy program found by GEPA

In [13]:
print(o.best_candidate["program"])

import dspy
import re
from typing import Optional

class MathQA(dspy.Signature):
    """
    Solve the given math word problem step by step, showing all necessary reasoning and calculations.
    Then, extract and return ONLY the final answer in the required format (number, variable, coordinate, interval, or list), with no units or extra text.

    Instructions:
    - Show clear, step-by-step reasoning in the 'reasoning' field.
    - In the 'answer' field, provide ONLY the final answer in the format required by the question:
        - If the answer is a number, return just the number (e.g., '42', '6', '0').
        - If the answer is a coordinate, return it in the form '(x,y)' (e.g., '(9,11)').
        - If the answer is a variable or object, return just its name (e.g., 'x', 'A').
        - If the answer is a list or set, return the elements separated by spaces (e.g., '2 3 5'), unless the question specifies a sum.
        - If the answer is an interval or union of intervals, use standar

Evaluating the optimized program

In [14]:
_ = adapter.evaluate(dataset.test, o.best_candidate)

2025/09/13 17:05:17 INFO dspy.evaluate.evaluate: Average Metric: 414.0 / 487 (85.0%)


We see it going from **67% to 93%** in just a few rounds of optimization!